src/image_surface_png.rs

```diff
@@ -14,61 +16,86 @@ use ImageSurface;
 
 struct ReadEnv<'a, R: 'a + Read> {
     reader: &'a mut R,
-    error: Option<Error>,
+    io_error: Option<Error>,
+    unwind_payload: Option<Box<dyn Any + Send + 'static>>,
 }
 
 unsafe extern "C" fn read_func<R: Read>(closure: *mut c_void, data: *mut u8, len: c_uint) -> cairo_status_t {
     let read_env: &mut ReadEnv<R> = &mut *(closure as *mut ReadEnv<R>);
     let buffer = slice::from_raw_parts_mut(data, len as usize);
-    match read_env.reader.read_exact(buffer) {
-        Ok(()) => Status::Success,
-        Err(error) => {
-            read_env.error = Some(error);
+    let result = std::panic::catch_unwind(AssertUnwindSafe(|| read_env.reader.read_exact(buffer)));
+    match result {
+        Ok(Ok(())) => {
+            Status::Success
+        }
+        Ok(Err(error)) => {
+            read_env.io_error = Some(error);
+            Status::ReadError
+        }
+        Err(payload) => {
+            read_env.unwind_payload = Some(payload);
             Status::ReadError
-        },
+        }
     }.into()
 }
 
 struct WriteEnv<'a, W: 'a + Write> {
     writer: &'a mut W,
-    error: Option<Error>,
+    io_error: Option<Error>,
+    unwind_payload: Option<Box<dyn Any + Send + 'static>>,
 }
 
 unsafe extern "C" fn write_func<W: Write>(closure: *mut c_void, data: *mut u8, len: c_uint) -> cairo_status_t {
     let write_env: &mut WriteEnv<W> = &mut *(closure as *mut WriteEnv<W>);
     let buffer = slice::from_raw_parts(data, len as usize);
-    match write_env.writer.write_all(buffer) {
-        Ok(()) => Status::Success,
-        Err(error) => {
-            write_env.error = Some(error);
+    let result = std::panic::catch_unwind(AssertUnwindSafe(|| write_env.writer.write_all(buffer)));
+    match result {
+        Ok(Ok(())) => {
+            Status::Success
+        }
+        Ok(Err(error)) => {
+            write_env.io_error = Some(error);
+            Status::WriteError
+        }
+        Err(payload) => {
+            write_env.unwind_payload = Some(payload);
             Status::WriteError
-        },
+        }
     }.into()
 }
 
 
 impl ImageSurface {
     pub fn create_from_png<R: Read>(stream: &mut R) -> Result<ImageSurface, IoError> {
```

Comment:
How is this safe, what ensures that the `&mut R` stays alive for as long as the `ImageSurface`?